homeassistant/helpers/floor_registry.py

```diff
@@ -0,0 +1,223 @@
+from typing import cast
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.util import slugify
+
+from . import area_registry as ar
+from .typing import UNDEFINED, UndefinedType
+
+DATA_REGISTRY = "floor_registry"
+EVENT_FLOOR_REGISTRY_UPDATED = "floor_registry_updated"
+STORAGE_KEY = "core.floor_registry"
+STORAGE_VERSION_MAJOR = 1
+SAVE_DELAY = 10
+
+
+@dataclass(slots=True, frozen=True)
+class FloorEntry:
+    """Floor registry entry."""
+
+    floor_id: str
+    name: str
+    normalized_name: str
+    icon: str | None = None
+
+
+class FloorRegistry:
+    """Class to hold a registry of floors."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the floor registry."""
+        self.hass = hass
+        self.floors: MutableMapping[str, FloorEntry] = {}
+        self._store = hass.helpers.storage.Store(
+            STORAGE_VERSION_MAJOR,
+            STORAGE_KEY,
+            atomic_writes=True,
+        )
+        self._normalized_name_floor_idx: dict[str, str] = {}
+        self.children: dict[str, set[str]] = {}
+
+    @callback
+    def async_get_floor(self, floor_id: str) -> FloorEntry | None:
+        """Get floor by id."""
+        return self.floors.get(floor_id)
+
+    @callback
+    def async_get_floor_by_name(self, name: str) -> FloorEntry | None:
+        """Get floor by name."""
+        normalized_name = normalize_floor_name(name)
+        if normalized_name not in self._normalized_name_floor_idx:
+            return None
+        return self.floors[self._normalized_name_floor_idx[normalized_name]]
+
+    @callback
+    def async_list_floors(self) -> Iterable[FloorEntry]:
+        """Get all floors."""
+        return self.floors.values()
+
+    @callback
+    def async_get_or_create(self, name: str) -> FloorEntry:
+        """Get or create an floor."""
+        if floor := self.async_get_floor_by_name(name):
+            return floor
+        return self.async_create(name)
+
+    @callback
+    def _generate_id(self, name: str) -> str:
+        """Initialize ID."""
+        suggestion = suggestion_base = slugify(name)
+        tries = 1
+        while suggestion in self.floors:
+            tries += 1
+            suggestion = f"{suggestion_base}_{tries}"
+        return suggestion
+
+    @callback
+    def async_create(
+        self,
+        name: str,
+        *,
+        icon: str | None = None,
+    ) -> FloorEntry:
+        """Create a new floor."""
+        normalized_name = normalize_floor_name(name)
+
+        if self.async_get_floor_by_name(name):
+            raise ValueError(f"The name {name} ({normalized_name}) is already in use")
+
+        floor = FloorEntry(
+            icon=icon,
+            floor_id=self._generate_id(name),
+            name=name,
+            normalized_name=normalized_name,
+        )
+        self.floors[floor.floor_id] = floor
+        self._normalized_name_floor_idx[normalized_name] = floor.floor_id
```

Comment:
Instead of multiple separate indices, make `FloorRegistry.floors` a `UserDict`, then the additional index is the responsability of the `UserDict` instead of spread out over the `FloorRegistry`. We use that pattern in the entity and device registries, for example `https://github.com/home-assistant/core/blob/2b4572bc6b55a53fc8fa5b9ed1185b25efa42a90/homeassistant/helpers/device_registry.py#L390`
We should use that pattern here and in the area registry too IMO.